src/transport/connection/tests.rs

```diff
@@ -0,0 +1,272 @@
+        responses: Vec<ResponseMessage>,
+    },
+    Request {
+        request: RequestMessage,
+        responses: Vec<ResponseMessage>,
+    },
+    // Subscription
+}
+
+#[derive(Clone, Debug)]
+enum Event {
+    Restart,
+    Message(Message),
+}
+
+impl Event {
+    fn request(request: &str, responses: &[&str]) -> Self {
+        let responses = responses
+            .into_iter()
+            .map(|s| ResponseMessage::from_simple(s))
+            .collect::<Vec<ResponseMessage>>();
+        let message = Message::Request {
+            request: RequestMessage::from_simple(request),
+            responses,
+        };
+        Event::Message(message)
+    }
+
+    fn handshake(request: &str, responses: &[&str]) -> Self {
+        let responses = responses
+            .into_iter()
+            .map(|s| ResponseMessage::from_simple(s))
+            .collect::<Vec<ResponseMessage>>();
+        let message = Message::Handshake {
+            request: RequestHandshake::from_simple(request),
+            responses,
+        };
+        Event::Message(message)
+    }
+}
+
+struct MockStream {
+    stream: TcpStream,
+    messages: VecDeque<Message>,
+    // keep the stream alive after messages have processed
+    keep_alive: bool,
+}
+
+impl MockStream {
+    pub fn new(stream: TcpStream, messages: VecDeque<Message>, keep_alive: bool) -> Self {
+        Self {
+            stream,
+            messages,
+            keep_alive,
+        }
+    }
+
+    fn read_prefix(&mut self) {
+        let mut data = vec![0_u8; 4];
+        self.stream.read(&mut data).unwrap();
+        assert_eq!(
+            data,
+            b"API\0",
+            "Handshake message did not start with API prefix.\n  left: {}\n  right: {}",
+            from_utf8(&data).unwrap(),
+            from_utf8(&b"API\0"[..]).unwrap(),
+        );
+    }
+
+    fn handle_message(&mut self, request: String, responses: Vec<ResponseMessage>) {
+        let message_size = read_header(&self.stream).unwrap();
+
+        let mut data = vec![0_u8; message_size];
+
+        self.stream.read_exact(&mut data).unwrap();
+
+        let raw_string = String::from_utf8(data).unwrap();
+        debug!("mock <- {:#?}", raw_string);
+
+        // let request = message.request.encode();
+
+        assert_eq!(
+            raw_string,
+            request,
+            // "Mock Server event request != Client request.\n  left: {},\n  right: {}",
+        );
+
+        for res in responses {
+            let packet = res.packet();
+            debug!("mock -> {:#?}", packet);
+            self.stream.write_all(packet.as_bytes()).unwrap();
+        }
+    }
+
+    pub fn process(&mut self) {
+        loop {
+            let message = match self.messages.pop_front() {
+                Some(message) => message,
+                None => break,
+            };
+
+            match message {
+                Message::Handshake { request, responses } => {
+                    self.read_prefix();
```

Comment:
Instead of a special message type, read_prefix will always be the first thing that the server does.